logic/resolvable/Concludable.java

```diff
@@ -233,27 +206,127 @@ boolean unificationSatisfiable(ThingVariable concludableThingVar, ThingVariable
+            };
+        } else if (value.isDateTime()) {
+            predicateFn = (a) -> {
+                if (!Encoding.ValueType.of(a.getType().getValueType().getValueClass())
+                        .comparableTo(Encoding.ValueType.DATETIME)) return false;
+                assert a.getType().isDateTime();
+                return a.asDateTime().getValue().compareTo(value.asDateTime().value()) == 0;
+            };
+        } else throw GraknException.of(ILLEGAL_STATE);
+        return predicateFn;
+    }
+
+    /** Relation handles these concludable patterns, where `$role` and `$relation` could be labelled, and there could
+     * be any number of rolePlayers:
+     * { $r($role: $x) isa $relation; }
+     * { ($role: $x) isa $relation; }
+     * { ($x) isa $relation; }
+     * { ($x); }
+     * { $r($x) isa $relation; }
+     * { $r($role: $x); }
+     * { $r($x); }
+     */
+    public static class Relation extends Concludable {
+
+        private final RelationConstraint relation;
+        private final IsaConstraint isa;
+        private final Set<LabelConstraint> labels;
+
+        private Relation(Conjunction conjunction, RelationConstraint relation, @Nullable IsaConstraint isa, Set<LabelConstraint> labels) {
+            super(conjunction);
+            this.relation = relation;
+            this.isa = isa;
+            this.labels = labels;
+        }
+
+        public static Relation of(RelationConstraint relation, @Nullable IsaConstraint isa, Set<LabelConstraint> labels) {
+            Conjunction.Cloner cloner;
+            IsaConstraint clonedIsa;
+            if (isa == null) {
+                cloner = Conjunction.Cloner.cloneExactly(labels, relation);
+                clonedIsa = null;
+            } else {
+                cloner = Conjunction.Cloner.cloneExactly(labels, isa, relation);
+                clonedIsa = cloner.getClone(isa).asThing().asIsa();
+            }
+            return new Relation(cloner.conjunction(), cloner.getClone(relation).asThing().asRelation(), clonedIsa,
+                                Iterators.iterate(labels).map(l -> cloner.getClone(l).asType().asLabel()).toSet());
+        }
+
+        public RelationConstraint relation() {
+            return relation;
+        }
 
-        public Relation(final RelationConstraint constraint, @Nullable IsaConstraint isaConstraint) {
-            super(ConstraintCopier.copyConstraint(constraint), isaConstraint == null ? set() : set(isaConstraint));
+        @Override
+        public Set<Constraint> concludableConstraints() {
+            return isa == null ? set(new HashSet<>(labels), relation) : set(new HashSet<>(labels), relation, isa);
         }
 
         @Override
+        ResourceIterator<Unifier> unify(Rule.Conclusion conclusion, ConceptManager conceptMgr) {
+            if (conclusion.isRelation()) return unify(conclusion.asRelation(), conceptMgr);
+            return Iterators.empty();
+        }
```

Comment:
This is now the entry point for unification. We could even skip this and go directly to the method below and do the casting there - thoughts? I quite like how obvious this check is right now though.